### PR TITLE
Use assembler to present products in cart

### LINKED ISSUE #20989: Product property hooks not available in cart (CartPresenter)
PR: #20980 

#### Describe the bug

The CartPresenter currently does not use the product assembler for Smart. This makes the available hooks for product properties not available in the cart.

#### Expected behavior

That the cart uses the same display/assemble logic as any other location in Prestashop.

Steps to reproduce the issue
1. Create a custom module using this `actionGetProductPropertiesAfter` hook
```php
    public function hookactionGetProductPropertiesAfter( $params )
    {
        $params['product']['test20989'] = 'Test property issue 20989';
    }
```
2. Edit a theme's cart template: `templates/checkout/_partials/cart-detailed-product-line.tpl`.
Add `{$product.test20989}` anywhere

3. Go to FO > Add product to cart > go checkout page
4. see error
-------------------------------------------

1. Add custom product properties using the `actionGetProductPropertiesAfter` hook.
2. They are currently not available in the cart templates.

#### Additional information

* PrestaShop version: develop / 1.7.7.x (tested locally on 1.7.6.x as well)
* PHP version: N/A

### PR DESCRIPTION
**Pull request by JoryHogeveen**
This allows other modules to use the `actionGetProductPropertiesAfter` hook in the cart as well.

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Support product property hooks for the cart presenter.
| Type?         | bug fix
| Category?     | FO
| Deprecations? | no
| Fixed ticket? | Fixes #20989 
| How to test?  | Add custom product properties using the `actionGetProductPropertiesAfter` hook. They are currently not available in the cart templates. With this PR they are.

Example hook function:
```php
    public function hookactionGetProductPropertiesAfter( $params )
    {
        $params['product']['test20989'] = 'Test property issue 20989';
    }
```
Next, go to your theme's cart template: `templates/checkout/_partials/cart-detailed-product-line.tpl`.
Add `{$product.test20989}` anywhere you want and with the PR applied it will output `Test property` issue #20989.
(Without the PR it will result in an error.)

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/20980)

### CODE REVIEW

**Review by PierreRambaud — DISMISSED** [review]
Could you target the develop branch and create an issue please?

**Comment by JoryHogeveen** [comment]
@PierreRambaud Done!

**Comment by Progi1984** [comment]
@JoryHogeveen Thanks. Could you fix PHPCSFixer & Travis CI ?

**Comment by JoryHogeveen** [comment]
@Progi1984 Done! 👍

File: src/Adapter/Presenter/Cart/CartPresenter.php [hunk 1]
```diff
@@ -95,6 +95,10 @@ private function includeTaxes()
      */
     private function presentProduct(array $rawProduct)
     {
+        $assembler = new \ProductAssembler(Context::getContext());
```
Comment (Progi1984) [inline src/Adapter/Presenter/Cart/CartPresenter.php]: Please set ProductAssembler in use parts.

Comment (Progi1984) [reply src/Adapter/Presenter/Cart/CartPresenter.php]: And it's :ok_hand: for me.

Comment (JoryHogeveen) [reply src/Adapter/Presenter/Cart/CartPresenter.php]: Done! 👍

**Review by Progi1984 — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Comment by PierreRambaud** [comment]
Hi @JoryHogeveen 
Could you provide us an example of the hook usage?

kind regards

**Comment by JoryHogeveen** [comment]
@PierreRambaud 
From issue:

Example hook function:

```php
    public function hookactionGetProductPropertiesAfter( $params )
    {
        $params['product']['test20989'] = 'Test property issue 20989';
    }
```

Next, go to your theme's cart template: `templates/checkout/_partials/cart-detailed-product-line.tpl`.
Add `{$product.test20989}` anywhere you want and with the PR applied it will output `Test property issue 20989`.
(Without the PR it will result in an error.)

**Comment by khouloudbelguith** [comment]
Hi @JoryHogeveen,

The exception is still displayed in the FO
image
image

I used this module
[mymod.zip](https://github.com/PrestaShop/PrestaShop/files/5238965/mymod.zip)
I edited this file
 `templates/checkout/_partials/cart-detailed-product-line.tpl` adding this line `{$product.test}`

Thanks to check and feedback.

**Comment by kpodemski** [comment]
@khouloudbelguith you have some additional checkings in Your module which may not allow Your module to assign new variable

I tested this change and for me it's all ok, maybe @JoryHogeveen could simplify code a bit:

```
$rawProduct = array_merge(
    (new ProductAssembler(Context::getContext()))->assembleProduct($rawProduct),
    $rawProduct
);
```

for me it's interesting that we need to have it directly in specific presenter 🤔 but changing this may impact performance so maybe it would be ok to leave this thought for now

It's working fine with both `actionGetProductProperties` hooks and also with whitelisted `Cart` properties 👍

**Comment by matks** [comment]
@kpodemski You're on fire tonight 😛 did you get a boost from [writing this great issue](https://github.com/PrestaShop/open-source/issues/44) 😉

**Comment by kpodemski** [comment]
@matks to be honest I'm miss those nights with digging deeply into PrestaShop core, back to the roots for me today ;)

**Comment by matks** [comment]
Night is indeed the very best time to focus on code 🌃

**Comment by JoryHogeveen** [comment]
@kpodemski 

Thanks for testing!

> maybe JoryHogeveen could simplify code a bit

Personally I find that more complex instead of simplified ;) but that is codestyle I guess. I'll leave it up to you guys!

**Comment by JoryHogeveen** [comment]
When is this PR planned for merge? Thanks!

**Comment by matks** [comment]
> When is this PR planned for merge? Thanks!

Oups! you did right to ping us, I think this PR got lost in state

So we should send it back to QA thanks to @kpodemski informations

**Comment by khouloudbelguith** [comment]
Hi @JoryHogeveen,

It is ok :heavy_check_mark: 
image
I used this module
[mymod.zip](https://github.com/PrestaShop/PrestaShop/files/5371873/mymod.zip)

Thanks!

**Comment by matks** [comment]
Thank you @JoryHogeveen

**Comment by JoryHogeveen** [comment]
Thanks everyone!

**Pull request merged**